## Day 16: Robust Model Evaluation

In this task, I implemented K-Fold Cross-Validation to evaluate the stability of my model instead of relying on a single train-test split.

I used a Random Forest Classifier and performed 5-Fold Cross-Validation to compute the mean accuracy and standard deviation.

Then, I compared training accuracy with validation accuracy to check whether the model is overfitting or generalizing well.

Finally, I performed a shuffle test using 10-Fold Cross-Validation to understand how shuffling affects model stability.

In [2]:
from sklearn.model_selection import KFold, cross_val_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import load_digits
import numpy as np

digits = load_digits()
X, y = digits.data, digits.target

model = RandomForestClassifier(n_estimators=50, random_state=42)

# Without shuffle
kf_no_shuffle = KFold(n_splits=10, shuffle=False)
scores_no_shuffle = cross_val_score(model, X, y, cv=kf_no_shuffle)

# With shuffle
kf_shuffle = KFold(n_splits=10, shuffle=True, random_state=42)
scores_shuffle = cross_val_score(model, X, y, cv=kf_shuffle)

print("No Shuffle STD:", np.std(scores_no_shuffle))
print("Shuffle STD:", np.std(scores_shuffle))

No Shuffle STD: 0.0229530777812618
Shuffle STD: 0.00929625412100634


Observation:

The standard deviation without shuffling is 0.0229, while with shuffling it reduces to 0.0093.

This shows a significant improvement in model stability after shuffling. Shuffling helps distribute the data more evenly across folds, reducing variation in performance and making the model more reliable.

Reflection:

If we do not shuffle the data in K-Fold, some folds may contain users mostly from one group, like Bangalore, while another fold may contain very few users from Delhi. This makes the training and testing distribution uneven. As a result, the model may learn patterns that work well for Bangalore users but fail for Delhi users. Shuffling helps mix the data more fairly across all folds, so the model is evaluated on a more representative sample of users.